nova/src/circuits/supernova/mod.rs

```diff
@@ -0,0 +1,631 @@
+                .get_or_insert_with(|| RelaxedR1CSInstance::<G1, C1>::new(&params.shapes[idx]));
+            let W_i = W[idx].get_or_insert_with(|| RelaxedR1CSWitness::zero(&params.shapes[idx]));
+
+            let U_secondary_i = U_secondary[idx]
+                .get_or_insert_with(|| RelaxedR1CSInstance::<G2, C2>::new(&params.shape_secondary));
+            let W_secondary_i = W_secondary[idx]
+                .get_or_insert_with(|| RelaxedR1CSWitness::zero(&params.shape_secondary));
+
+            let proof = NIMFSProof::<G1, G2, C1, C2, RO>::prove(
+                &params.pp,
+                &params.pp_secondary,
+                &params.ro_config,
+                &params.digest,
+                (&params.shapes[idx], &params.shape_secondary),
+                (U_i, W_i),
+                (U_secondary_i, W_secondary_i),
+                (&u, &w),
+            )?;
+
+            let input =
+                SuperNovaAugmentedCircuitInput::NonBase(SuperNovaAugmentedCircuitNonBaseInput {
+                    vk: params.digest,
+                    i: G1::ScalarField::from(i),
+                    pc_prev: pc,
+                    pc: pc_next,
+                    z_0: z_0.clone(),
+                    z_i,
+                    U: U.clone(),
+                    U_secondary: U_secondary.clone(),
+                    u,
+                    proof: proof.0,
+                });
+
+            let (U_p, W_p) = proof.1;
+            let (U_secondary_p, W_secondary_p) = proof.2;
+
+            U[idx].replace(U_p);
+            W[idx].replace(W_p);
+
+            U_secondary[idx].replace(U_secondary_p);
+            W_secondary[idx].replace(W_secondary_p);
+
+            let i_next = i.saturating_add(1);
+
+            (i_next, pc_next, input, U, W, U_secondary, W_secondary)
+        } else {
+            let pc = step_circuit.pc_init();
+            let num_augmented_circuits = SC::NUM_CIRCUITS;
+
+            let U = vec![None; num_augmented_circuits];
+            let U_secondary = vec![None; num_augmented_circuits];
+
+            let W = vec![None; num_augmented_circuits];
+            let W_secondary = vec![None; num_augmented_circuits];
+
+            let input = SuperNovaAugmentedCircuitInput::<G1, G2, C1, C2, RO>::Base {
+                vk: params.digest,
+                z_0: z_0.clone(),
+                pc,
+                num_augmented_circuits,
+            };
+            let i_next = 1;
+
+            (i_next, pc, input, U, W, U_secondary, W_secondary)
+        };
+
+        let cs = ConstraintSystem::new_ref();
+        cs.set_mode(SynthesisMode::Prove { construct_matrices: false });
+
+        let circuit = SuperNovaAugmentedCircuit::new(&params.ro_config, step_circuit, input);
+
+        let (pc_next, z_i) = tracing::debug_span!(target: LOG_TARGET, "satisfying_assignment")
+            .in_scope(|| {
+                SuperNovaConstraintSynthesizer::generate_constraints(circuit, cs.clone())
+            })?;
+
```

Comment:
```suggestion
        
        let (pc_next, z_i) = tracing::debug_span!(target: LOG_TARGET, "satisfying_assignment")
            .in_scope(|| {
                SuperNovaConstraintSynthesizer::generate_constraints(circuit, cs.clone())
            })?;

        let circuit = SuperNovaAugmentedCircuit::new(&params.ro_config, pc_next, step_circuit, input);
```
we can just pass the result of computing the `pc` here, so it doesn't need to be in the `input`.